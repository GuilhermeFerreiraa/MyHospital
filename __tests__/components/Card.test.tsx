import Card from '@/components/card';
import { fireEvent, render } from '@testing-library/react-native';
import { useRouter as useRouterOriginal } from 'expo-router';
import React from 'react';
import { Alert } from 'react-native';
import renderer from 'react-test-renderer';

jest.mock('@/components/toast', () => ({
	ToastMessage: {
		update: jest.fn(),
	},
}));

jest.mock('react-native-toast-notifications', () => ({
	useToast: jest.fn(() => ({
	})),
}));

jest.mock('expo-router', () => ({
	useRouter: jest.fn(),
}));

const mockAddCandidacy = jest.fn();
const mockRemoveCandidacy = jest.fn();
jest.mock('@/stores/candidacy-store', () => ({
	useCandidacyStore: () => ({
		addCandidacy: mockAddCandidacy,
		removeCandidacy: mockRemoveCandidacy,
	}),
}));

jest.spyOn(Alert, 'alert').mockImplementation((title, message, buttons) => {
	if (buttons && buttons.length > 1 && buttons[1].onPress) {
		// Simulate clicking 'Sim'
		buttons[1].onPress();
	}
});

beforeEach(() => {
	jest.useFakeTimers();
});

afterEach(() => {
	jest.useRealTimers();
});

const mockJobData = {
	id: 'fdb298a4-9ddb-42df-86ec-fc1641f68dc7',
	created_at: '2024-07-16T22:07:33.270Z',
	role: 'Enfermeiro',
	requirement: 'Experiência de no mínimo 1 ano em clínicas.',
	schedule: {
		starts_at: '2025-04-18T19:53:19.302Z',
		ends_at: '2025-04-19T07:53:19.302Z',
	},
	shift_type: 'Noturno',
	candidacy_sent: false,
};

describe('Card Component', () => {
	test('renders correctly with given data', () => {
		const tree = renderer.create(<Card data={mockJobData} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	test('calls handleApplyJob when apply button is pressed', () => {
		const { getByTestId } = render(<Card data={mockJobData} />);

		fireEvent.press(getByTestId('apply-button'));

		jest.runAllTimers();
		expect(mockAddCandidacy).toHaveBeenCalledWith(mockJobData);
	});

	test('calls handleCancelAplicationJob when cancel button is pressed', () => {
		const mockRouterBack = jest.fn();

		(useRouterOriginal as jest.Mock).mockReturnValue({ back: mockRouterBack });

		const { getByTestId } = render(<Card data={{ ...mockJobData, candidacy_sent: true }} />);

		fireEvent.press(getByTestId('cancel-button'));

		jest.runAllTimers();
		expect(mockRemoveCandidacy).toHaveBeenCalledWith(mockJobData.id);
		expect(mockRouterBack).toHaveBeenCalled();
	});
});
