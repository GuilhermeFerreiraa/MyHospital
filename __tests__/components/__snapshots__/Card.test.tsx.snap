// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders correctly with given data 1`] = `
<View
  collapsable={false}
  jestAnimatedStyle={
    {
      "value": {},
    }
  }
  style={
    {
      "backgroundColor": "#fff",
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "marginBottom": 8,
      "marginTop": 8,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
      "width": "100%",
    }
  }
>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "alignItems": "center",
        },
        {
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          {
            "fontFamily": "Inter_600SemiBold",
          },
          {
            "color": "#2E4A71",
          },
        ]
      }
    >
      Enfermeiro
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginLeft": -8,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 14,
            "lineHeight": 20,
          },
          {
            "fontFamily": "Inter_500Medium",
          },
          {
            "color": "#2E4A71",
          },
          [
            {
              "marginLeft": 8,
            },
          ],
        ]
      }
    >
      Data:
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "lineHeight": 20,
          },
          {
            "fontFamily": "Inter_400Regular",
          },
          {
            "color": "#374151",
          },
          [
            {
              "marginLeft": 8,
            },
          ],
        ]
      }
    >
      18 de abr. de 2025
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginLeft": -8,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 14,
            "lineHeight": 20,
          },
          {
            "fontFamily": "Inter_500Medium",
          },
          {
            "color": "#2E4A71",
          },
          [
            {
              "marginLeft": 8,
            },
          ],
        ]
      }
    >
      Horário:
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "lineHeight": 20,
          },
          {
            "fontFamily": "Inter_400Regular",
          },
          {
            "color": "#374151",
          },
          [
            {
              "marginLeft": 8,
            },
          ],
        ]
      }
    >
      16:53h
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginLeft": -8,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 14,
            "lineHeight": 20,
          },
          {
            "fontFamily": "Inter_500Medium",
          },
          {
            "color": "#2E4A71",
          },
          [
            {
              "marginLeft": 8,
            },
          ],
        ]
      }
    >
      Turno:
    </Text>
    <Text
      style={
        [
          {
            "fontSize": 14,
            "lineHeight": 20,
          },
          {
            "fontFamily": "Inter_400Regular",
          },
          {
            "color": "#374151",
          },
          [
            {
              "marginLeft": 8,
            },
          ],
        ]
      }
    >
      Noturno
    </Text>
  </View>
  <Text
    style={
      [
        {
          "fontSize": 14,
          "lineHeight": 20,
        },
        {
          "fontFamily": "Inter_400Regular",
        },
        {
          "color": "#374151",
        },
      ]
    }
  >
    Experiência de no mínimo 1 ano em clínicas.
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2E4A71",
        "borderBottomLeftRadius": 6,
        "borderBottomRightRadius": 6,
        "borderTopLeftRadius": 6,
        "borderTopRightRadius": 6,
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "center",
        "marginTop": 16,
        "opacity": 1,
        "width": "100%",
      }
    }
    testID="apply-button"
  >
    <Text />
    <Text
      style={
        [
          {
            "color": "#fff",
          },
          {
            "fontFamily": "Inter_600SemiBold",
          },
          {
            "textAlign": "center",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          {
            "marginLeft": 8,
            "marginRight": 8,
          },
        ]
      }
    >
      Candidatar-me
    </Text>
  </View>
</View>
`;
